subsys/net/ip/net_pkt.c

```diff
@@ -1654,6 +1654,59 @@ void net_pkt_get_info(struct k_mem_slab **rx,
 	}
 }
 
+int net_pkt_get_src_addr(struct net_pkt *pkt, struct sockaddr *addr,
+			 socklen_t addrlen)
+{
+	struct net_context *ctx;
+	enum net_ip_protocol proto;
+	sa_family_t family;
+	u16_t port = 0;
+
+	if (!addr || !pkt) {
+		return -EINVAL;
+	}
+
+	ctx = pkt->context;
+	proto = net_context_get_ip_proto(ctx);
+	family = net_context_get_family(ctx);
+
+	addr->sa_family = family;
+
+	if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
+		struct net_tcp_hdr *hdr = net_pkt_tcp_data(pkt);
+
+		port = hdr->src_port;
+	} else if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
+		struct net_udp_hdr *hdr = net_pkt_udp_data(pkt);
+
+		port = hdr->src_port;
+	} else {
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
```

Comment:
If you move this block before checking UDP/TCP, then you could add the proto resolving here.